### PR TITLE
Support building PersistentIndex from pre-existing tablet data

### LINKED ISSUE #3950: Support building PersistentIndex from pre-existing tablet data
When BE is upgraded from versions without PersistentIndex support, it needs to be built from the tablet's primary keys.
And in some scenarios, the index may be corrupt due to bugs or other issues, so rebuilding from original data is also a requirement.

### PR DESCRIPTION
**Pull request by sevev**
support building PersistentIndex from pre-existing tablet data

## What type of PR is this：
- [ ] bug
- [ ] feature
- [ ] enhancement
- [x] others

## Which issues of this PR fixes ：

Fixes #3950

## Problem Summary(Required) ：

PersistentIndex should be built from the tablet's rowset data; there are two conditions:
+ BE is upgraded from versions without PersistentIndex support
+ In some scenarios, the index may be corrupt due to bugs or other issues, so rebuilding from original data is also required.

### CODE REVIEW

File: be/src/storage/persistent_index.cpp [hunk 1]
```diff
@@ -1082,6 +1090,222 @@ Status PersistentIndex::_load(const PersistentIndexMetaPB& index_meta) {
...
+    }
+    RETURN_IF_ERROR(_insert_rowsets(tablet, rowsets, pkey_schema, apply_version, std::move(pk_column)));
+    if (size() != total_rows - total_dels) {
+        LOG(WARNING) << strings::Substitute("load primary index row count not match tablet:$0 index:$1 != stats:$2",
+                                            tablet->tablet_id(), size(), total_rows - total_dels);
+    }
+    RETURN_IF_ERROR(_build_commit(tablet, index_meta));
+    LOG(INFO) << "build persistent index finish tablet: " << tablet->tablet_id() << " version:" << apply_version
+              << " #rowset:" << rowsets.size() << " #segment:" << total_segments << " data_size:" << total_data_size
+              << " cost time: " << timer.elapsed_time() / 1000000 << "ms";
```
Comment (decster) [inline be/src/storage/persistent_index.cpp:1305]: cost time: -> time

File: be/src/storage/persistent_index.cpp [hunk 2]
```diff
@@ -1082,6 +1090,222 @@ Status PersistentIndex::_load(const PersistentIndexMetaPB& index_meta) {
...
+        std::string l0_index_file_path_tmp = l0_index_file_path + ".tmp";
+        fs::CreateBlockOptions wblock_opts({l0_index_file_path});
+        wblock_opts.mode = Env::MUST_EXIST;
+        if (_dump_snapshot) {
+            RETURN_IF_ERROR(_block_mgr->create_block(wblock_opts, &_index_block));
+        } else {
+            RETURN_IF_ERROR(Env::Default()->rename_file(l0_index_file_path_tmp, l0_index_file_path));
+            RETURN_IF_ERROR(_block_mgr->create_block(wblock_opts, &_index_block));
+        }
+    }
```
Comment (decster) [inline be/src/storage/persistent_index.cpp:1120]: Is there a better way to maintain _index_block? Maybe you can only build index in-memory without writing WAL(snapshot is always better than WAL in this case), and only do snapshot/flush.

**Review by decster — CHANGES_REQUESTED** [review]

File: be/src/storage/persistent_index.cpp [hunk 3]
```diff
@@ -1082,6 +1090,222 @@ Status PersistentIndex::_load(const PersistentIndexMetaPB& index_meta) {
...
+    // we can load from index file directly
+    EditVersion lastest_applied_version;
+    tablet->updates()->get_latest_applied_version(&lastest_applied_version);
+    RETURN_IF_ERROR(create(fix_size, lastest_applied_version));
+    if (status.ok()) {
+        // all applied rowsets has save in existing persistent index meta
+        // so we can load persistent index according to PersistentIndexMetaPB
+        EditVersion version = index_meta.version();
+        if (version == lastest_applied_version) {
+            status = load(index_meta);
```
[the remark this answers is not in the document]
Comment (sevev) [reply be/src/storage/persistent_index.cpp]: > If do load, do not do create, if do create do not do load, only build pkey_schema & fix_size if do create. So should check versions match first.

done, change in https://github.com/StarRocks/starrocks/pull/4377/commits/1b7de4cb120d74c24d0fcac8e5019d528b870983

File: be/src/storage/persistent_index.cpp [hunk 4]
```diff
@@ -1082,6 +1090,222 @@ Status PersistentIndex::_load(const PersistentIndexMetaPB& index_meta) {
...
+        return Status::NotSupported("Only PrimaryKey table is supported to use persistent index");
+    }
+
+    const TabletSchema& tablet_schema = tablet->tablet_schema();
+    vector<ColumnId> pk_columns(tablet_schema.num_key_columns());
+    for (auto i = 0; i < tablet_schema.num_key_columns(); i++) {
+        pk_columns[i] = (ColumnId)i;
+    }
+    auto pkey_schema = vectorized::ChunkHelper::convert_schema_to_format_v2(tablet_schema, pk_columns);
+    size_t fix_size = PrimaryKeyEncoder::get_encoded_fixed_size(pkey_schema);
```
[the remark this answers is not in the document]
Comment (sevev) [reply be/src/storage/persistent_index.cpp]: done, change in https://github.com/StarRocks/starrocks/pull/4377/commits/1b7de4cb120d74c24d0fcac8e5019d528b870983

File: be/src/storage/persistent_index.cpp (hunk 2 above)
Comment (sevev) [reply be/src/storage/persistent_index.cpp]: done, change in https://github.com/StarRocks/starrocks/pull/4377/commits/1b7de4cb120d74c24d0fcac8e5019d528b870983

**Review by decster — APPROVED** [review]

**Review by chaoyli — APPROVED** [review]

**Pull request merged**